File: mobile/lib/domain/entities/community.dart
import '../../core/utils/format.dart';
import 'user.dart';

class Post {
  Post({
    required this.id,
    required this.author,
    required this.title,
    this.content,
    this.imageUrl,
    this.category,
    this.diagnosisId,
    this.likes = 0,
    this.comments = 0,
    this.likedByMe = false,
    this.createdAt,
  });
  final String id, title;
  final AppUser author;
  final String? content, imageUrl, category, diagnosisId;
  final int likes, comments;
  final bool likedByMe;
  final DateTime? createdAt;

  factory Post.fromJson(Map<String, dynamic> j) => Post(
        id: j['id'],
        author: AppUser.fromJson(j['author']),
        title: j['title'],
        content: j['content'],
        imageUrl: j['image_url'],
        category: j['category'],
        diagnosisId: j['diagnosis_id'],
        likes: j['likes_count'] ?? 0,
        comments: j['comments_count'] ?? 0,
        likedByMe: j['liked_by_me'] ?? false,
        createdAt: parseDate(j['created_at']),
      );
}

class Comment {
  Comment({required this.id, required this.author, required this.content, this.createdAt});
  final String id, content;
  final AppUser author;
  final DateTime? createdAt;

  factory Comment.fromJson(Map<String, dynamic> j) =>
      Comment(id: j['id'], author: AppUser.fromJson(j['author']), content: j['content'], createdAt: parseDate(j['created_at']));
}

File: backend/app/schemas/community.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.common import ORMModel
from app.schemas.user import UserPublic

CATEGORY = r"^(disease|question|experience|advice)$"


class PostIn(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    content: str | None = None
    image_url: str | None = None
    category: str | None = Field(default=None, pattern=CATEGORY)
    diagnosis_id: uuid.UUID | None = None


class PostUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    content: str | None = None
    image_url: str | None = None
    category: str | None = Field(default=None, pattern=CATEGORY)


class PostOut(ORMModel):
    id: uuid.UUID
    author: UserPublic
    diagnosis_id: uuid.UUID | None = None
    title: str
    content: str | None = None
    image_url: str | None = None
    category: str | None = None
    likes_count: int = 0
    comments_count: int = 0
    liked_by_me: bool = False
    created_at: datetime | None = None


class CommentIn(BaseModel):
    content: str = Field(min_length=1, max_length=2000)


class CommentOut(ORMModel):
    id: uuid.UUID
    post_id: uuid.UUID
    author: UserPublic
    content: str
    created_at: datetime | None = None


class ReportIn(BaseModel):
    post_id: uuid.UUID | None = None
    comment_id: uuid.UUID | None = None
    reason: str = Field(max_length=100)
    description: str | None = None


class ReportOut(ORMModel):
    id: uuid.UUID
    reporter_id: uuid.UUID
    post_id: uuid.UUID | None = None
    comment_id: uuid.UUID | None = None
    reason: str | None = None
    description: str | None = None
    status: str
    created_at: datetime | None = None
    reviewed_at: datetime | None = None
